### ISSUE TITLE
Hotkey settings JS needs improvement

### ISSUE THREAD

**Issue by MightyPork** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#413]
I currently don't have time to rewrite it, but it sorely needs it.

`inject/settings/hotkeys.js`

- can't detect anything outside letters, arrows, ctrl, alt and shift
- -> Super key not supported (Meta on Linux, Cmd on Mac)
- -> AltGr, Menu not supported
- -> Fx don't work, etc etc
- can't set hotkey without CTRL
- can't have ALT and SHIFT together
- multimedia keys don't work (although this should be enabled by default - then again, what if users want to change it ?)
- the code is a huge mess

note: Manually editing the config file works, so there's that for the time being.

reference: https://github.com/atom/electron/blob/master/docs/api/accelerator.md

**Comment by MarshallOfSound** [comment]
Yeah, this is literally the worst file ever....  I wrote the stupidist shortcut parser in a few minutes and I'm regretting it now :+1:

**Comment by baleksey** [comment]
Can somebody give me a hint how and where to modify hotkeys config files to setup my keys (especially Fs: F10, F11..)? As MightyPork wrote it's unusable for me at them moment. thank you!

**Comment by MightyPork** [comment]
If you locate the settings file, which is called '.settings.json', you can
edit it there.

(But I think you have to first set the hotkey in the GUI for it to be
visible in the json.)

If you're on Windows, I think the file will be in %APPDATA% somewhere, look
for a folder called "GPMDP_STORE" or "Google Play Music Desktop
Player/json_store" if you run it from source.
On Mar 10, 2016 8:58 AM, "baleksey" <notifications@github.com> wrote:

Can somebody give me a hint how and where to modify hotkeys config files to
setup my keys (especially Fs: F10, F11..)? As MightyPork wrote it's
unusable for me at them moment. thank you!

—
Reply to this email directly or view it on GitHub
<https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/413#issuecomment-194721539>
.

**Comment by baleksey** [comment]
MightyPork, thanks a lot! It works.

**Comment by natas72716** [comment]
I am having the same issue on ubuntu 15.10, 
I am unable to enter any keys through the GUI.
Will try editing the config manually and see if that helps.

**Comment by SicariusNoctis** [comment]
Perhaps using a [library](http://stackoverflow.com/questions/2353417/which-is-the-best-javascript-keyboard-event-library-hotkeys-shortcuts) might make things easier?

**Comment by toddmedema** [comment]
Would also be great to add support for the spacebar key in Windows - https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/950

**Comment by MarshallOfSound** [comment]
I have some code that fixes this issue in a **really** (IMO) nice way, but it is currently blocked on OS X by this issue

https://github.com/electron/electron/issues/5188

Just giving you guys an update

**Comment by ccolorado** [comment]
Not sure if this should be a separate issue. But the Hotkeys Section can't be edited at all on my archlinux system. 
No matter what combinations, or how many times I hit the escape key. The  "Not Set" text never changes.

**Comment by MarshallOfSound** [comment]
@ccolorado Have you tried a simple "Control + Shift + P" combination.  Currently all combinations require a control key

**Comment by ccolorado** [comment]
@MarshallOfSound Actually I was trying Ctrl+F10, Ctrl+F11, Ctrl+F12 for previous, play/pause, next.

Ctrl+Shif+P worked tho. I see issue [430](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/413) actually reports this.

**Comment by glorygeek** [comment]
Not sure if same issue, but I have Ctr + Alt + G as a hotkey outside of GPM, and now when I try to start, it crashes and gives this error window 
![image](https://cloud.githubusercontent.com/assets/3461135/15526733/f151a822-21e8-11e6-9a96-72c33527bd8e.png)

**Comment by MarshallOfSound** [comment]
@glorygeek Completely unrelated issue :laughing: 

Simply go to `%APPDATA%\Google Play Music Desktop Player\json_store`

And delete the `'.settings.json'` file

Then the app will work again

**Comment by glorygeek** [comment]
Thanks!

**Comment by raulpinpla** [comment]
@jostrander Hi Jostrander I post here to maintain the other thread closed, 
The version is 3.2.5 and I download de .deb from the offical website and install with  dpkg .

**Comment by baleksey** [comment]
Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: Samuel Attard <notifications@github.com> Date: 5/24/16  22:54  (GMT-05:00) To: MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL- <Google-Play-Music-Desktop-Player-UNOFFICIAL-@noreply.github.com> Cc: baleksey <baleksey@gmail.com>, Comment <comment@noreply.github.com> Subject: Re: [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-] Hotkey settings JS needs improvement (#413) 
@glorygeek Completely unrelated issue 😆 

Simply go to %APPDATA%\Google Play Music Desktop Player\json_store

And delete the '.settings.json' file

Then the app will work again

**Comment by MarshallOfSound** [comment]
I'm going to close this out as "Can't Fix" due to how macOS send key events (in particular the meta key)
